### PR TITLE
Feature signing service message oriented protocol

### PR DESCRIPTION
**Pull request by rwrzesien**
Resolves https://github.com/golemfactory/concent/issues/625

Running sample command:
```bash
python -m signing_service.signing_service 127.0.0.1 "85cZzVjahnRpUBwm0zlNnqTdYom1LF1P1WNShLg17cmhN2UssnPrCjHKTi5susO3wrr/q07eswumbL82b4HgOw==" --signing-service-private-key "GfkWzCRGL/Me2wk41mS0NRSch2XMpgIXYLRaasmcozw=" --ethereum-private-key "MQ=="
```

### CODE REVIEW

File: signing_service/signing_service/tests/tests.py [hunk 1]
```diff
@@ -107,14 +118,68 @@ def test_that_signing_service_should_reraise_different_socket_erros(self):
...
+class SigningServiceHandleConnectionTestCase(TestCase):
+
+    def setUp(self):
+        self.host = '127.0.0.1'
+        self.port = 8000
+        self.initial_reconnect_delay = 2
+        # TODO: Use predefined ones when SigningService will accept Concent private key as parameter.
+        self.concent_public_key = b'\xf3\x97\x19\xcdX\xda\x86tiP\x1c&\xd39M\x9e\xa4\xddb\x89\xb5,]O\xd5cR\x84\xb85\xed\xc9\xa17e,\xb2s\xeb\n1\xcaN.l\xba\xc3\xb7\xc2\xba\xff\xabN\xde\xb3\x0b\xa6l\xbf6o\x81\xe0;'
+
+    def test_that__handle_connection_should_send_wrapped_service_refuse_if_frame_is_invalid(self):
```
Comment (kbeker) [inline signing_service/signing_service/tests/tests.py]: Too much bottom highlights  `that__handle`

File: signing_service/signing_service/tests/tests.py [hunk 2]
```diff
@@ -107,14 +118,68 @@ def test_that_signing_service_should_reraise_different_socket_erros(self):
...
 
 
+class SigningServiceHandleConnectionTestCase(TestCase):
+
+    def setUp(self):
+        self.host = '127.0.0.1'
+        self.port = 8000
+        self.initial_reconnect_delay = 2
+        # TODO: Use predefined ones when SigningService will accept Concent private key as parameter.
+        self.concent_public_key = b'\xf3\x97\x19\xcdX\xda\x86tiP\x1c&\xd39M\x9e\xa4\xddb\x89\xb5,]O\xd5cR\x84\xb85\xed\xc9\xa17e,\xb2s\xeb\n1\xcaN.l\xba\xc3\xb7\xc2\xba\xff\xabN\xde\xb3\x0b\xa6l\xbf6o\x81\xe0;'
```
Comment (kbeker) [inline signing_service/signing_service/tests/tests.py]: Please use `CONCENT_PUBLIC_KEY` which you generate in the top of file

File: signing_service/signing_service/tests/tests.py [hunk 3]
```diff
@@ -107,14 +118,68 @@ def test_that_signing_service_should_reraise_different_socket_erros(self):
...
+
+    def test_that__handle_connection_should_send_wrapped_service_refuse_if_frame_is_invalid(self):
+        # Prepare message with wrong signature
+        middleman_message = GolemMessageMiddlemanMessage(Ping(), 99)
+        raw_message = middleman_message.serialize(
+            # TODO: Use predefined ones when SigningService will accept Concent private key as parameter.
+            private_key=b'l\xcdh\x19\xeb$>\xbcG\xa1\xc7v\xe8\xd7o\x0c\xbf\x0e\x0fM\x89lw\x1e\xd7K\xd6Hnv$\xa2',
```
Comment (kbeker) [inline signing_service/signing_service/tests/tests.py]: Also please use `CONCENT_PRIVATE_KEY` which you generated on the top of this file

File: signing_service/signing_service/tests/tests.py (hunk 1 above)
Comment (rwrzesien) [reply signing_service/signing_service/tests/tests.py]: This is the name of the method, it starts with `_`.

File: signing_service/signing_service/tests/tests.py (hunk 2 above)
Comment (rwrzesien) [reply signing_service/signing_service/tests/tests.py]: Please read the comment above this line ;)

File: signing_service/signing_service/tests/tests.py (hunk 3 above)
Comment (rwrzesien) [reply signing_service/signing_service/tests/tests.py]: Same here.

**Comment by Jakub89** [comment]
I'm having a hard time running signing_service script from this branch. Command like this:
``` shell
./signing_service.sh http://127.0.0.1 1 b'\xf3\x97\x19\xcdX\xda\x86tiP\x1c&\xd39M\x9e\xa4\xddb\x89\xb5,]O\xd5cR\x84\xb85\xed\xc9\xa17e,\xb2s\xeb\n1\xcaN.l\xba\xc3\xb7\xc2\xba\xff\xabN\xde\xb3\x0b\xa6l\xbf6o\x81\xe0;'
```
raise error:
``` shell
raise SigningServiceValidationError('concent_public_key is not valid public key.')
signing_service.exceptions.SigningServiceValidationError: concent_public_key is not valid public key.
```

**Comment by rwrzesien** [comment]
@Jakub89 You can't pass bytes through command line. It will get changed to string and malformed along the way. You have to do it like:

`./signing_service.sh http://127.0.0.1 1 "\xf3\x97\x19\xcdX\xda\x86tiP\x1c&\xd39M\x9e\xa4\xddb\x89\xb5,]O\xd5cR\x84\xb85\xed\xc9\xa17e,\xb2s\xeb\n1\xcaN.l\xba\xc3\xb7\xc2\xba\xff\xabN\xde\xb3\x0b\xa6l\xbf6o\x81\xe0;"`

File: signing_service/signing_service/signing_service.py [hunk 4]
```diff
@@ -136,6 +180,11 @@ def _parse_arguments() -> argparse.Namespace:
         type=int,
         help='Initial delay between reconnections, doubles after each unsuccessful attempt and is reset after success.',
     )
+    parser.add_argument(
+        'concent_public_key',
+        type=str,
```
Comment (Jakub89) [inline signing_service/signing_service/signing_service.py]: It would be nice to have some short version of arguments, like `-c` for that `--concent-cluster-port`, and similar for others parameters.

Comment (rwrzesien) [reply signing_service/signing_service/signing_service.py]: :+1: done

File: signing_service/signing_service/signing_service.py [hunk 5]
```diff
@@ -105,9 +113,37 @@ def _connect(self, tcp_socket: socket.socket) -> None:
 
     def _handle_connection(self, tcp_socket: socket.socket) -> None:  # pylint: disable=no-self-use
         """ Inner loop that handles data exchange over socket. """
-        while True:
-            data = tcp_socket.recv(1024)
-            tcp_socket.send(data)
+        while not self._was_sigterm_caught():
+            raw_message_received = next(handle_frame_receive(tcp_socket, separator_escaping_expected=True))
+            try:
+                _middleman_message_payload = AbstractMiddlemanMessage.deserialize(
```
[the remark this answers is not in the document]
Comment (rwrzesien) [reply signing_service/signing_service/signing_service.py]: It is done this way according to new design.

**Comment by cameel** [comment]
@rwrzesien @Jakub89 Please also see comments in #671. Please talk with @kbeker and decide in which format the key is going to be passed to the service. You could use the Python string representation you showed above or encode it, e.g. with base64.

Either way is fine (though I'd be wary of bash escaping if you decide to use a Python string) but please choose one, because I think that currently @kbeker has implemented it differently.

File: signing_service/signing_service/signing_service.py [hunk 6]
```diff
@@ -186,13 +186,15 @@ def _parse_arguments() -> argparse.Namespace:
         help="Concent's public key.",
     )
     parser.add_argument(
+        '-p',
```
Comment (cameel) [inline signing_service/signing_service/signing_service.py]: This is not really related to the protocol. It should be in a separate commit. Or even a separate pull request because this could go to master immediately, without further review.

File: signing_service/signing_service/signing_service.py [hunk 7]
```diff
@@ -105,9 +113,35 @@ def _connect(self, tcp_socket: socket.socket) -> None:
...
+                deserialized_message = next(
+                    receive_frame(
+                        connection=tcp_socket,
+                        public_key=b'\xf3\x97\x19\xcdX\xda\x86tiP\x1c&\xd39M\x9e\xa4\xddb\x89\xb5,]O\xd5cR\x84\xb85\xed\xc9\xa17e,\xb2s\xeb\n1\xcaN.l\xba\xc3\xb7\xc2\xba\xff\xabN\xde\xb3\x0b\xa6l\xbf6o\x81\xe0;',
+                    )
+                )
+                if isinstance(deserialized_message.payload, Message):
+                    deserialized_message.payload.verify_signature(
+                        public_key=b'\xf3\x97\x19\xcdX\xda\x86tiP\x1c&\xd39M\x9e\xa4\xddb\x89\xb5,]O\xd5cR\x84\xb85\xed\xc9\xa17e,\xb2s\xeb\n1\xcaN.l\xba\xc3\xb7\xc2\xba\xff\xabN\xde\xb3\x0b\xa6l\xbf6o\x81\xe0;',
+                    )
```
Comment (cameel) [inline signing_service/signing_service/signing_service.py]: Doesn't Middleman protocol already check the signature? It looks like something that belongs in there. It knows that it's dealing with golem-messages after all.

File: signing_service/signing_service/signing_service.py [hunk 8]
```diff
@@ -105,9 +113,35 @@ def _connect(self, tcp_socket: socket.socket) -> None:
...
+            except (InvalidSignature, MiddlemanProtocolError) as exception:
+                logger.error(f'Exception caught when deserializing received Middleman protocol message: {exception}.')
```
Comment (cameel) [inline signing_service/signing_service/signing_service.py]: This is not our error - it's invalid input from the client. It should be logged at `INFO` level just like validation errors in Concent views.

File: signing_service/signing_service/signing_service.py [hunk 9]
```diff
@@ -105,9 +113,35 @@ def _connect(self, tcp_socket: socket.socket) -> None:
...
+                service_refused = ServiceRefused()
```
Comment (cameel) [inline signing_service/signing_service/signing_service.py]: Please send an `ERROR` frame instead.

#618 originally said to use `ServiceRefused` but I forgot to update the description after I have added `ERROR` frames to the blueprint.

File: signing_service/signing_service/tests/tests.py [hunk 10]
```diff
@@ -13,13 +21,17 @@
...
+
+
 class SigningServiceMainTestCase(TestCase):
 
     def setUp(self):
         self.host = '127.0.0.1'
         self.port = 8000
         self.initial_reconnect_delay = 2
-        self.concent_public_key = b'\xf3\x97\x19\xcdX\xda\x86tiP\x1c&\xd39M\x9e\xa4\xddb\x89\xb5,]O\xd5cR\x84\xb85\xed\xc9\xa17e,\xb2s\xeb\n1\xcaN.l\xba\xc3\xb7\xc2\xba\xff\xabN\xde\xb3\x0b\xa6l\xbf6o\x81\xe0;'
+        self.concent_public_key = CONCENT_PUBLIC_KEY
```
Comment (cameel) [inline signing_service/signing_service/tests/tests.py]: Please don't hard-code keys. I see that this was already hard-coded in one of the previous issues so we can't really remove it though. Next time, if there's anything preventing you from not hard-coding it, please tell me. I think this was hard-coded because I forgot to add the command-line parameter for it in issue description.

File: signing_service/signing_service/tests/tests.py (hunk 2 above)
Comment (cameel) [reply signing_service/signing_service/tests/tests.py]: I have updated issue description yesterday. Please add command-line options for this in this pull request. And please remove all hard-coded keys from the code. We rarely ever want to have keys hard-coded so if you see no other choice, it's very likely that it's my mistake.

File: signing_service/signing_service/tests/tests.py (hunk 10 above)
Comment (rwrzesien) [reply signing_service/signing_service/tests/tests.py]: It is not in master yet but in pull request which I wanted to merge and on which this one is based, so I will replace it there.

**Comment by rwrzesien** [comment]
@cameel @Jakub89 @kbeker 

> @rwrzesien @Jakub89 Please also see comments in #671. Please talk with @kbeker and decide in which format the key is going to be passed to the service. You could use the Python string representation you showed above or encode it, e.g. with base64.

Lets make it `base64` then, I see @kbeker used it, and I have already updated https://github.com/golemfactory/concent/pull/680 and merged into master.

**Comment by rwrzesien** [comment]
@kbeker No, I don't want you to change, I have agreed that we should go this way. So if @Jakub89 agrees we are all clear.

**Comment by Jakub89** [comment]
@cameel @rwrzesien @kbeker 

> @rwrzesien @Jakub89 Please also see comments in #671. Please talk with @kbeker and decide in which format the key is going to be passed to the service. You could use the Python string representation you showed above or encode it, e.g. with base64.

This doesn't really effects me, so I guess `base64` is fine.

File: signing_service/signing_service/signing_service.py (hunk 8 above)
Comment (rwrzesien) [reply signing_service/signing_service/signing_service.py]: :+1: done

File: signing_service/signing_service/tests/tests.py (hunk 2 above)
Comment (rwrzesien) [reply signing_service/signing_service/tests/tests.py]: :+1: done

File: signing_service/signing_service/signing_service.py (hunk 6 above)
Comment (rwrzesien) [reply signing_service/signing_service/signing_service.py]: :+1: done

File: signing_service/signing_service/signing_service.py [hunk 11]
```diff
@@ -35,17 +44,29 @@ class SigningService:
...
+        port: int,
+        initial_reconnect_delay: int,
+        concent_public_key: bytes,
+        signing_service_private_key: bytes,
+    ) -> None:
         assert isinstance(host, str)
         assert isinstance(port, int)
         assert isinstance(initial_reconnect_delay, int)
         assert isinstance(concent_public_key, bytes)
+        assert isinstance(signing_service_private_key, bytes)
```
Comment (Jakub89) [inline signing_service/signing_service/signing_service.py]: There seems to be no default value for `signing_service_private_key`, and there should be unless it's not optional argument.
```
./signing_service.sh 127.0.0.1 'yCDaoFX/lgp26x88USJbe8OCkLfkRPcZ8CqjX1cj4tatObo2qVjAa51Mimyurp1kLWj3Z6HqOXY0wxxsagmn4Q=='
```
```
Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/jakub/Documents/work/projects/concent/signing_service/signing_service/signing_service.py", line 334, in <module>
    arg_signing_service_private_key,
  File "/home/jakub/Documents/work/projects/concent/signing_service/signing_service/signing_service.py", line 64, in __init__
    assert isinstance(signing_service_private_key, bytes)
AssertionError
```

Comment (rwrzesien) [reply signing_service/signing_service/signing_service.py]: @cameel @kbeker This is caused by setting default on arguments which can be retrieved in 3 ways (console, file, env), and not adding `required=True` to group definition. By default code is trying to get variable from env, and when it does not exist, it fails. I suggest removing default and adding `required=True` to group definition if you both agree.

Comment (cameel) [reply signing_service/signing_service/signing_service.py]: Sounds fine to me. The option specifying that we should get the value from a variable will still be optional in that case, right?

File: signing_service/signing_service/signing_service.py [hunk 12]
```diff
@@ -211,10 +255,40 @@ def __call__(self, parser, namespace, values, option_string=None):
...
+        help='Singing Service private key.',
+    )
+    signing_service_private_key_parser_group.add_argument(
+        '--signing-service-private-key-from-env',
+        dest='signing_service_private_key',
+        action=make_secret_provider_factory(env_variable_name='SIGNING_SERVICE_PRIVATE_KEY', base64_convert=True),
+        help='Singing Service private key.',
+    )
+    parser.set_defaults(
+        signing_service_private_key=b64decode(os.environ.get('SIGNING_SERVICE_PRIVATE_KEY'))  # type: ignore
```
Comment (dybi) [inline signing_service/signing_service/signing_service.py]: hm... conditional for default doesn't seem to be a good idea...

File: signing_service/signing_service/tests/tests.py [hunk 13]
```diff
@@ -116,6 +139,67 @@ def test_that_signing_service_should_reraise_different_socket_erros(self):
...
 
 
+class SigningServiceHandleConnectionTestCase(TestCase):
+
+    def setUp(self):
+        self.host = '127.0.0.1'
+        self.port = 8000
+        self.initial_reconnect_delay = 2
+
+    def test_that__handle_connection_should_send_wrapped_service_refuse_if_frame_is_invalid(self):
```
Comment (dybi) [inline signing_service/signing_service/tests/tests.py]: where is `ServiceRefused` in this test?

File: signing_service/signing_service/tests/tests.py [hunk 14]
```diff
@@ -116,6 +139,67 @@ def test_that_signing_service_should_reraise_different_socket_erros(self):
...
+        # Prepare message with wrong signature
+        middleman_message = GolemMessageFrame(Ping(), 99)
+        raw_message = append_frame_separator(
+            escape_encode_raw_message(
+                middleman_message.serialize(
+                    private_key=CONCENT_PRIVATE_KEY,
+                )
+            )
+        )
+        malformed_raw_message = bytes(bytearray([max(raw_message[1] - 1, 0)])) + raw_message[1:]
```
Comment (dybi) [inline signing_service/signing_service/tests/tests.py]: In case that `raw_message[0]==0` and `raw_message[1]==0`, `raw_message == malformed_message`, so test might be not deterministic

File: signing_service/signing_service/tests/tests.py [hunk 15]
```diff
@@ -116,6 +139,67 @@ def test_that_signing_service_should_reraise_different_socket_erros(self):
...
+                        signing_service = SigningService(
+                            self.host,
+                            self.port,
+                            self.initial_reconnect_delay,
+                            CONCENT_PUBLIC_KEY,
+                            SIGNING_SERVICE_PRIVATE_KEY,
+                        )
+
+                        # For test purposes we reverse roles, so signing service works as server.
+                        signing_service_socket.bind(('127.0.0.1', 8001))
```
Comment (dybi) [inline signing_service/signing_service/tests/tests.py]: this is a bit risky... please use `unused_tcp_port` fixture (https://github.com/pytest-dev/pytest-asyncio#unused_tcp_port)

File: signing_service/signing_service/tests/tests.py [hunk 16]
```diff
@@ -21,6 +30,9 @@
 concent_ecc_keys = ECCx(None)
 (CONCENT_PRIVATE_KEY, CONCENT_PUBLIC_KEY) = concent_ecc_keys.raw_privkey, concent_ecc_keys.raw_pubkey
 
+signing_service_ecc_keys = ECCx(None)
+(SIGNING_SERVICE_PRIVATE_KEY, SIGNING_SERVICE_PUBLIC_KEY) = signing_service_ecc_keys.raw_privkey, signing_service_ecc_keys.raw_pubkey
+
 
 class SigningServiceMainTestCase(TestCase):
```
Comment (dybi) [inline signing_service/signing_service/tests/tests.py]: quite frankly, here as well `unused_tcp_port` should be used (basically, everywhere where a real server is being used)

File: signing_service/signing_service/signing_service.py (hunk 11 above)
Comment (rwrzesien) [reply signing_service/signing_service/signing_service.py]: It will be `required` to chose exactly one of three ways of passing variable (console, file, env variable).

File: signing_service/signing_service/signing_service.py (hunk 12 above)
Comment (rwrzesien) [reply signing_service/signing_service/signing_service.py]: Removes as discussed in https://github.com/golemfactory/concent/pull/688#discussion_r208591838.

File: signing_service/signing_service/signing_service.py [hunk 17]
```diff
@@ -137,6 +150,9 @@ def _validate_arguments(self):
 def _parse_arguments() -> argparse.Namespace:
     def make_secret_provider_factory(read_command_line=False, env_variable_name=None, use_file=False, base64_convert=False):
         def wrapper(**kwargs):
+            # Remove 'required' as it is added automatically by ArgumentParser.
+            if 'required' in kwargs:
+                del kwargs['required']
```
[the remark this answers is not in the document]
Comment (rwrzesien) [reply signing_service/signing_service/signing_service.py]: It is in separated 971b5b0e5bdffc4c47f9ea020aad1866261c7b87.

File: signing_service/signing_service/signing_service.py [hunk 18]
```diff
@@ -119,9 +127,28 @@ def _connect(self, tcp_socket: socket.socket) -> None:
...
+        for raw_message_received in receive_frame_generator:
+            try:
+                AbstractFrame.deserialize(
+                    raw_message=raw_message_received,
+                    public_key=self.signing_service_public_key,
+                )
+            except (InvalidSignature, MiddlemanProtocolError) as exception:
+                logger.info(f'Exception caught when deserializing received Middleman protocol message: {exception}.')
+                middleman_message_response = ErrorFrame(
+                    payload=(ErrorCode.InvalidFrameSignature, str(exception)),
```
[the remark this answers is not in the document]
Comment (rwrzesien) [reply signing_service/signing_service/signing_service.py]: :+1: done

File: signing_service/signing_service/utils.py [hunk 19]
```diff
@@ -12,3 +16,64 @@ def is_valid_public_key(key):
...
+        dest,
+        required=False,
+        help=None  # pylint: disable=redefined-builtin
+    ):
+        self.read_command_line = read_command_line
+        self.env_variable_name = env_variable_name
+        self.use_file = use_file
+        self.base64_convert = base64_convert
+
+        command_line_arguments = 0 if self.env_variable_name is not None else None
```
[the remark this answers is not in the document]
Comment (rwrzesien) [reply signing_service/signing_service/utils.py]: It can also be none,  so `_count` can be misleading too, I have removed separate variable for it completely and moved the condition to where its result is used directly.

File: signing_service/signing_service/signing_service.py [hunk 20]
```diff
@@ -211,10 +227,39 @@ def __call__(self, parser, namespace, values, option_string=None):
...
+        ethereum_private_key=b64decode(os.environ.get('ETHEREUM_PRIVATE_KEY'))  # type: ignore
+        if os.environ.get('ETHEREUM_PRIVATE_KEY') is not None
+        else None
+    )
+
+    signing_service_private_key_parser_group = parser.add_mutually_exclusive_group()
+    signing_service_private_key_parser_group.add_argument(
+        '--signing-service-private-key',
+        dest='signing_service_private_key',
+        action=make_secret_provider_factory(read_command_line=True, base64_convert=True),
```
[the remark this answers is not in the document]
Comment (rwrzesien) [reply signing_service/signing_service/signing_service.py]: But they are already configured in different way, so to make them generic there has to be a some of extra code with parameters, so I would skip this now, it is very readable in the current way.

File: signing_service/signing_service/signing_service.py [hunk 21]
```diff
@@ -106,9 +127,28 @@ def _connect(self, tcp_socket: socket.socket) -> None:
...
+            try:
+                AbstractFrame.deserialize(
+                    raw_message=raw_message_received,
+                    public_key=self.signing_service_public_key,
+                )
+            except (InvalidSignature, MiddlemanProtocolError) as exception:
+                logger.info(f'Malformed messages failed to deserialize with exception: {exception}.')
+                middleman_message_response = ErrorFrame(
+                    payload=(ErrorCode.InvalidFrameSignature, str(exception)),
+                    request_id=99,
```
Comment (dybi) [inline signing_service/signing_service/signing_service.py]: I would suggest making it `0` (restricted `request_id`) and extracting to a (protocol) constant

Comment (rwrzesien) [reply signing_service/signing_service/signing_service.py]: :+1: done
